service/storage_service.cc

```diff
@@ -1453,11 +1453,22 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
             group0_guard guard, const raft_topology_cmd& cmd,
             const std::unordered_set<raft::server_id>& exclude_nodes,
             drop_guard_and_retake drop_and_retake = drop_guard_and_retake::yes) {
-        auto nodes = _topo_sm._topology.normal_nodes
-            | boost::adaptors::filtered([&exclude_nodes] (const std::pair<const raft::server_id, replica_state>& n) {
-                return !exclude_nodes.contains(n.first);
-            })
-            | boost::adaptors::map_keys;
+        std::vector<raft::server_id> nodes;
+        nodes.reserve(_topo_sm._topology.normal_nodes.size() + _topo_sm._topology.transition_nodes.size());
+        for (const auto& node: _topo_sm._topology.normal_nodes) {
+            if (!exclude_nodes.contains(node.first)) {
+                nodes.push_back(node.first);
+            }
+        }
+
+        // We must send barrier_and_drain to the decommissioning node as it might be coordinating requests.
+        auto it = _topo_sm._topology.transition_nodes.begin();
+        auto node_is_decommissioning = it != _topo_sm._topology.transition_nodes.end()
+                && (it->second.state == node_state::decommissioning || it->second.state == node_state::left_token_ring);
```

Comment:
The problem is that failed bootstrap and replace also go through the `node_state::left_token_ring`. Not sure we want to send the barrier to then though especially as they may be dead by now.

Reply:
We probably shouldn't send the barrier to them. I'll investigate what exactly happens and think about a fix if needed.

Reply:
The following happens after we start the rollback of a bootstrap or replace and the joining node is dead:
- the `left_token_ring` state handler in `handle_node_transition` calls `global_token_metadata_barrier`,
- `global_token_metadata_barrier` sends `barrier_and_drain`, which fails because it includes the dead joining node,
- `global_token_metadata_barrier` catches the exception, proceeds and sends `barrier`, which behaves the same way before and after the change because the joining node is not included.

So, IIUC, the `left_token_ring` state handler for a failed bootstrap or decommission should succeed after the change if it would succeed before the change. We will log an error after failed `barrier_and_drain`, but we are on the error path, so some errors have already been logged. In conclusion, I don't think this change will have a noticeable bad impact.

To fix this, we need a way to distinguish a finishing decommission from a failed bootstrap or replace. The only idea I have is introducing a new state for a failed bootstrap and replace.

I opened #16871. This is a small issue, but the bottom part of if -- "Another issue is in `handle_node_transition`" -- is another reason why we need a way to distinguish these two cases.

I can introduce this state in this PR or as a follow-up. It should be easy and add only a few lines of code. I can also implement a different solution if someone has an idea. cc @kbr-scylla

For now, I'll do fixes from the other comments and send a new version.

Reply:
I have nothing against introducing a new state. And if we want to introduce a new state we should do it before branching. @gleb-cloudius please give your opinion.

Reply:
> The only idea I have is introducing a new state for a failed bootstrap and replace.

Another is to make `left_token_ring` to be a transition state, not a node state. @kbr-scylla do you remember what is the reason it is not?

Reply:
```
commit b8ddfd9ef977d074fbc49fc8becac6e651b95a81
Author: Kamil Braun <kbraun@scylladb.com>
Date:   Fri May 19 17:30:43 2023 +0200

    raft topology: introduce `left_token_ring` state
    
    We want for the decommissioning node to wait before shutting down until
    every node learns that it left the token ring. Otherwise some nodes may
    still try coordinating writes to that nodes after it already shut down,
    leading to unnecessary failures on the data path(e.g. for CL=ALL writes).
    
    Before this change, a node would shut down immediately after observing
    that it was in `left` state; some other nodes may still see it in
    `decommissioning` state and the topology transition state as
    `write_both_read_new`, so they'd try to write to that node.
    
    After this change, the node first enters the `left_token_ring` state
    before entering `left`, while the topology transition state is removed
    (so we've finished the token ring change - the node no longer has tokens
    in the ring, but it's still part of the topology). There we perform a
    read barrier, allowing all nodes to observe that the decommissioning
    node has indeed left the token ring. Only after that barrier succeeds we
    allow the node to shut down.

```

The purpose of left_token_ring was to remove the node's tokens from the token ring before having it shut down. I guess it felt more natural to use node_state than transition_state to decide whether a node should keep its tokens. Perhaps combining `node_state == decommissioning && transition_state == left_token_ring` would achieve the same thing. And then we could execute different logic for `node_state == bootstrapping && transition_state == left_token_ring` (this would happen during rollback), etc.

Reply:
> And then we could execute different logic for node_state == bootstrapping && transition_state == left_token_ring (this would happen during rollback), etc.

Yes. Today we lose info about what operation is ongoing when we move to `left_token_ring`. Introducing yet another `left_token_ring_after_rollback` is a way to preserve it, but making `left_token_ring` a transition state feels more correct.

Reply:
Then I'll make `left_token_ring` a transition state. Should I do it in this PR or can we leave it as an immediate follow-up? Leaving this as a follow-up would allow us to stop ignoring stale topology exceptions in tests earlier.

Reply:
IMO follow-up

Reply:
> IMO follow-up

OK